### PR TITLE
SSR-2261: Attachment

### PR DESCRIPTION
**Pull request by h000780**
## Checklist

- [ ] Read CONTRIBUTING.md([for v1](https://github.com/kintone-labs/kintone-ui-component/master/CONTRIBUTING.md), [for v0](https://github.com/kintone-labs/kintone-ui-component/v0_dev/CONTRIBUTING.md))
- [ ] Updated documentation if it is required.
- [ ] Added tests if it is required.
- [ ] Passed lint and test command.

### CODE REVIEW

File: src/attachment/index.ts [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+        case "zh":
+          return zh;
+        case "ja":
+          return ja;
+        default:
+          return en;
+      }
+    }
+
+    private _handleClickFileRemove(event: any) {
```
Comment (voduchau) [inline src/attachment/index.ts:202]: Change any to PointerEvent:
private _handleClickFileRemove(event: PointerEvent) {
      const removeButtonEl = event.currentTarget as HTMLButtonElement;
      const index = parseInt(
        removeButtonEl.getAttribute("data-file-index")!,
        10
      );

File: src/attachment/index.ts [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+      const index = parseInt(
+        event.currentTarget.getAttribute("data-file-index"),
+        10
+      );
+      if (!isNaN(index) && this.files) {
```
Comment (voduchau) [inline src/attachment/index.ts:207]: do we need check isNaN ?

File: src/attachment/index.ts [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+        return Math.round(size / ATTACHMENT.UNIT_LENGTH.ONE_GB) + " GB";
+      } else if (size >= ATTACHMENT.UNIT_LENGTH.ONE_MB) {
+        return Math.round(size / ATTACHMENT.UNIT_LENGTH.ONE_MB) + " MB";
+      } else if (size >= ATTACHMENT.UNIT_LENGTH.ONE_KB) {
+        return Math.round(size / ATTACHMENT.UNIT_LENGTH.ONE_KB) + " KB";
+      }
+      return Math.round(size) + " bytes";
+    }
+
+    private _isFileOrDirectoryDrag = (event: DragEvent) => {
```
Comment (voduchau) [inline src/attachment/index.ts:327]: private _isFileOrDirectoryDrag = (event: DragEvent) => {
      if (!event.dataTransfer) return false;

      if (event.dataTransfer.items !== undefined) {
        for (let i = 0; i < event.dataTransfer.items.length; i++) {
          if (event.dataTransfer.items[i].kind.toLowerCase() === "file") {
            return true;
          }
        }
      }

      if (event.dataTransfer.types !== undefined) {
        for (let i = 0; i < event.dataTransfer.types.length; i++) {
          if (event.dataTransfer.types[i].toLowerCase() === "files") {
            return true;
          }
        }
      }
      return false;
    };

File: src/attachment/index.ts [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        this.requestUpdate();
+      }
+    }
+
+    private _handleDragEnter(event: DragEvent) {
+      this._dragEnterCounter++;
+      if (this._dragEnterCounter === 1 && this._isFileOrDirectoryDrag(event)) {
+        event.preventDefault();
+
+        this._groupFilesEl.style.height =
```
Comment (voduchau) [inline src/attachment/index.ts:228]: Move these magic numbers in _handleDragEnter function to constant variable.
16, 4, 6

File: src/base/kuc-base.ts [hunk 5]
```diff
@@ -1,12 +1,18 @@
...
 import { v4 as uuid } from "uuid";
+import { FileItem } from "../attachment/type";
 
 type CustomEventDetail = {
   data?: string | null;
   oldValue?: string | string[];
   value?: string | string[];
   error?: string;
 };
+type AttachmentEventDetail = {
```
Comment (voduchau) [inline src/base/kuc-base.ts:11]: the property of AttachmentEventDetail is correct ?
I don't see any file using changedFiles.
Please check it

File: src/base/attachment/resource/constant.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+export const ATTACHMENT = {
+  UNIT_LENGTH: {
+    ONE_KB: 1024,
+    ONE_MB: 1048576,
+    ONE_GB: 1073741824,
+  },
+  INVALID_SIZE_ERROR: "Nan size",
```
Comment (voduchau) [inline src/base/attachment/resource/constant.ts:7]: I think it should be:
export const ONE_KB = 1024;
export const ONE_MB = ONE_KB ^ 2;
export const ONE_GB = ONE_KB ^ 3;
export const ATTACHMENT_INVALID_SIZE_ERROR = "NaN size";

File: src/attachment/index.ts [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+            this._getAttachmentItemTemplete(item, number)
+          )}
+          </ul>
+          <div class="kuc-attachment__group__files__browse-button"
+          ?hidden="${this._isDraging || this.disabled}">
+            <span class="kuc-attachment__group__files__browse-button__text">${
+              this._locale.ATTACHMENT_BROWSE
+            }</span>
+            <div class="kuc-attachment__group__files__browse-button__input-container">
+              <input class="kuc-attachment__group__files__browse-button__input-container__input" type="file" accept multiple
```
Comment (voduchau) [inline src/attachment/index.ts:115]: about accept attribute in <input >
why it is empty ? do we need add it ?

File: src/attachment/index.ts [hunk 8]
```diff
@@ -62,7 +76,15 @@ let exportAttachment;
       const validProps = validateProps(props);
       Object.assign(this, validProps);
     }
-
+    shouldUpdate(changedProperties: PropertyValues): boolean {
+      if (changedProperties.has("files")) {
+        if (!isArrayType<FileItem>(this.files)) {
```
Comment (voduchau) [inline src/attachment/index.ts:82]: you can combine the condition:
if (
        changedProperties.has("files") &&
        !isArrayType<FileItem>(this.files)
      ) {
        throwErrorAfterUpdateComplete(this, ERROR_MESSAGE.FILES.IS_NOT_ARRAY);
        return false;
      }

File: src/attachment/index.ts [hunk 9]
```diff
@@ -310,21 +340,24 @@ let exportAttachment;
...
+        return Math.round(size / ONE_KB) + " KB";
       }
       return Math.round(size) + " bytes";
     }
 
     private _isFileOrDirectoryDrag = (event: DragEvent) => {
+      if (!event.dataTransfer) {
+        return false;
+      }
       if (event.dataTransfer && event.dataTransfer.items !== undefined) {
```
Comment (voduchau) [inline src/attachment/index.ts:373]: remove event.dataTransfer
because we check !event.dataTransfer above.

File: src/attachment/type.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  | File
+  | { name: string; size: string; [key: string]: any };
+export declare type AttachmentProps = {
+  className?: string;
+  disabled?: boolean;
+  error?: string;
+  files?: FileItem[];
+  id?: string;
+  label?: string;
+  language?: "ja" | "en" | "zh" | "auto";
```
Comment (voduchau) [inline src/attachment/type.ts:14]: add zh-TW to language

File: src/attachment/index.stories.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    actions: {
+      handles: ["change"],
+    },
+  },
+};
+const Template = (args) => {
+  const handleAttachmentChange = (event) => {
+    console.log(event);
+  };
+  return html`
```
Comment (voduchau) [inline src/attachment/index.stories.js:33]: change to new way storybook:const Template = (args) => {
  const attachment = new Attachment({ ...args });
  attachment.addEventListener("change", (event) => {
    console.log(event, "event");
  });
  return attachment;
};

File: src/attachment/index.ts [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+            </button>
+          </div>
+          <span class="kuc-attachment__group__files__display-area__item__size">
+            ${this._getFileSize(item.size)}
+          </span>
+        </li>
+      `;
+    }
+    private _getRemoveButtonIcon() {
+      return html`<svg
```
Comment (voduchau) [inline src/attachment/index.ts:187]: import { ... svg } from "lit";
..
return svg`<svg>.....`
*replace html by svg

File: src/attachment/test/changeEvent.test.ts [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+      inputEl.dispatchEvent(new Event("change"));
+      await elementUpdated(container);
+      expect(triggeredEvent.type).to.equal("change");
+      expect(triggeredEvent.detail.files[0].name).to.equal("filename.jpg");
+      expect(triggeredEvent.detail.type).to.equal("add-file");
+    });
+
+    it("should be triggered add file event when drag file into drag file area ", async () => {
+      let triggeredEvent: any = null;
+      const container = new Attachment({ disabled: true });
```
Comment (voduchau) [inline src/attachment/test/changeEvent.test.ts:29]: why disabled: = true but you can catches "add-files" event ?

File: src/attachment/test/files.test.ts [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+const files = [
+  { name: "file.txt", size: "150" },
+  { name: "icon.jpg", size: "200" },
+];
+const invalidFileSizeFiles = [
+  { name: "file.txt", size: "text" },
+  { name: "icon.jpg", size: "200" },
+];
+describe("Attachment", () => {
+  describe("value", () => {
```
Comment (voduchau) [inline src/attachment/test/files.test.ts:13]: describe("files") not "value"

File: src/attachment/test/files.test.ts [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    it("should be none checked items when not assinged on constructor", async () => {
```
Comment (voduchau) [inline src/attachment/test/files.test.ts:14]: I think "should be none checked items when not assinged on constructor" not correct.
Please change it

File: src/attachment/test/files.test.ts [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+      const container = new Attachment({ files: files });
+      const el = await fixture(container);
+      const lisEl = el.querySelectorAll(
+        ".kuc-attachment__group__files__display-area__item__name"
+      );
+      expect(lisEl.length).to.equal(files.length);
+      lisEl.forEach((liEl, index) => {
```
Comment (voduchau) [inline src/attachment/test/files.test.ts:21]: please don't write logic in unit test.
we will expect directly
expect((lisEl[0] as HTMLDivElement).innerText).to.equal(files[0].name);
expect((lisEl[1] as HTMLDivElement).innerText).to.equal(files[1].name);

File: src/attachment/test/files.test.ts [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+      const container = new Attachment({
+        files: files,
+      });
+      container.files = invalidFileSizeFiles;
+      const el = await fixture(container);
+      const lisEl = el.querySelectorAll(
+        ".kuc-attachment__group__files__display-area__item__size"
+      );
+      expect(lisEl.length).to.equal(files.length);
+      lisEl.forEach((liEl, index) => {
```
Comment (voduchau) [inline src/attachment/test/files.test.ts:36]: please don't write logic in unit test.
we will expect directly
example comment above

File: src/base/attachment/resource/constant.ts [hunk 18]
```diff
@@ -0,0 +1,4 @@
+export const ONE_KB = 1024;
+export const ONE_MB = 1048576;
+export const ONE_GB = 1073741824;
+export const ATTACHMENT_INVALID_SIZE_ERROR = "Nan size";
```
Comment (voduchau) [inline src/base/attachment/resource/constant.ts:4]: "NaN size" is better, right ?

File: src/attachment/type.ts [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+  disabled?: boolean;
+  error?: string;
+  files?: FileItem[];
+  id?: string;
+  label?: string;
+  language?: "ja" | "en" | "zh" | "zh-TW" | "auto";
+  requiredIcon?: boolean;
+  visible?: boolean;
+};
+export declare class Button extends KucBase {
```
Comment (voduchau) [inline src/attachment/type.ts:18]: change "export declare class Button" to "export declare class Attachment"

File: src/attachment/index.ts [hunk 20]
```diff
@@ -189,6 +201,38 @@ let exportAttachment;
         </li>
       `;
     }
+    async updated(_changedProperties: PropertyValues) {
+      await this.updateComplete;
+      this._updateFileNameMaxWidth();
+    }
+
+    private _updateFileNameMaxWidth() {
```
Comment (dong0) [inline src/attachment/index.ts:209]: You can change the CSS file to adjust the width of root.
And only adjust the width of file name in this function.

**Pull request merged**